{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f56b12cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6883e541",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot_df = pd.read_csv('../data/annotation/cannonical_annotation.csv',sep = '\\t',header = 0,index_col = 0)\n",
    "transcript_df = pd.read_csv('../data/annotation/annot_label.csv',sep = '\\t',header = 0)\n",
    "transcript_df['length'] = np.absolute(transcript_df['Start']-transcript_df['End'])\n",
    "#How many sequences if we fully chop transcriptome\n",
    "np.sum(np.ceil(transcript_df['length']/512))\n",
    "annot_df = annot_df[annot_df['Annotation'].isin(['exon','UTR',])]\n",
    "annot_df['GeneID']=pd.Categorical(annot_df['GeneID'],transcript_df['Gene'])\n",
    "annot_df = annot_df.sort_values(by=['Chrom','GeneID','Start']).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b1b4a62e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reverse_strand = {'+':'-','-':'+'}\n",
    "length = 512\n",
    "chrom = []\n",
    "strand = []\n",
    "start_l = []\n",
    "end_l = []\n",
    "gene = []\n",
    "for index,row in transcript_df.iterrows():\n",
    "    start = row['Start']\n",
    "    end = row['End']\n",
    "    split = [(round(length*i)+start, round(length*(i+1))+start) for i in range(int(math.ceil((end-start)/length)))]\n",
    "    for entry in split:\n",
    "        chrom.append(row['Chr'])\n",
    "        strand.append(reverse_strand[row['Strand']])\n",
    "        start_l.append(entry[0])\n",
    "        end_l.append(entry[1])\n",
    "        gene.append(row['Gene'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85212861",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_df = pd.DataFrame({'Chr':chrom,'Start':start_l,'End':end_l,'Strand':strand,'Gene':gene})\n",
    "bed_df['Gene']=pd.Categorical(bed_df['Gene'],transcript_df['Gene'])\n",
    "bed_df = bed_df.sort_values(by=['Chr','Gene','Start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9448d977",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_pointer = 0\n",
    "annot_pointer = 0\n",
    "label = np.empty(len(bed_df), dtype=object)\n",
    "label[...] = [set()for _ in range(label.shape[0])]\n",
    "#while bed_df.iloc[bed_pointer]['Gene'] == annot_df.iloc[annot_pointer]['GeneID']:\n",
    "while bed_pointer < len(bed_df)-1  or annot_pointer < len(annot_df)-1:\n",
    "    bed_row = bed_df.iloc[bed_pointer]\n",
    "    annot_row = annot_df.iloc[annot_pointer]\n",
    "    #chromosome check:\n",
    "    if bed_row['Chr'] < annot_row['Chrom']:\n",
    "        bed_pointer += 1\n",
    "        continue\n",
    "    elif bed_row['Chr'] > annot_row['Chrom']:\n",
    "        annot_pointer += 1\n",
    "        continue\n",
    "    #not overlapping， move pointer \n",
    "    if bed_row['End'] <= annot_row['Start']:\n",
    "        bed_pointer += 1\n",
    "        continue\n",
    "    elif bed_row['Start'] >= annot_row['End']:\n",
    "        annot_pointer += 1\n",
    "        continue\n",
    "    #overlapping\n",
    "    if max(bed_row['Start'], annot_row['Start']) <= min(bed_row['End'], annot_row['End']):\n",
    "        #complete overlap:\n",
    "        if (bed_row['Start'] >= annot_row['Start']) and (bed_row['End'] <= annot_row['End']):\n",
    "            label[bed_pointer].add(annot_row['Annotation'])\n",
    "            bed_pointer += 1 \n",
    "        #partial overlap\n",
    "        elif bed_row['Start'] <= annot_row['Start']:\n",
    "            label[bed_pointer].update(['junction',annot_row['Annotation']])\n",
    "            bed_pointer += 1\n",
    "        #partial overlap\n",
    "        elif bed_row['End'] >= annot_row['End']:\n",
    "            label[bed_pointer].update([annot_row['Annotation'],'junction'])\n",
    "            annot_pointer += 1\n",
    "               "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4da9318b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>12381</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>junction,exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12381</td>\n",
       "      <td>12893</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>junction,exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12893</td>\n",
       "      <td>13405</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>junction,exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13405</td>\n",
       "      <td>13917</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13917</td>\n",
       "      <td>14429</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>junction,exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819693</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57212696</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819694</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57212696</td>\n",
       "      <td>57213208</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>junction,exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819695</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57213208</td>\n",
       "      <td>57213720</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>junction,exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819696</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57213720</td>\n",
       "      <td>57214232</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>junction,exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819697</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57214232</td>\n",
       "      <td>57214744</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3819756 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Chr     Start       End Strand                     Gene  \\\n",
       "0        chr1     11869     12381      -        ENSG00000223972.5   \n",
       "1        chr1     12381     12893      -        ENSG00000223972.5   \n",
       "2        chr1     12893     13405      -        ENSG00000223972.5   \n",
       "3        chr1     13405     13917      -        ENSG00000223972.5   \n",
       "4        chr1     13917     14429      -        ENSG00000223972.5   \n",
       "...       ...       ...       ...    ...                      ...   \n",
       "3819693  chrY  57212184  57212696      +  ENSG00000227159.8_PAR_Y   \n",
       "3819694  chrY  57212696  57213208      +  ENSG00000227159.8_PAR_Y   \n",
       "3819695  chrY  57213208  57213720      +  ENSG00000227159.8_PAR_Y   \n",
       "3819696  chrY  57213720  57214232      +  ENSG00000227159.8_PAR_Y   \n",
       "3819697  chrY  57214232  57214744      +  ENSG00000227159.8_PAR_Y   \n",
       "\n",
       "                 Label  \n",
       "0        junction,exon  \n",
       "1        junction,exon  \n",
       "2        junction,exon  \n",
       "3                 exon  \n",
       "4        junction,exon  \n",
       "...                ...  \n",
       "3819693           exon  \n",
       "3819694  junction,exon  \n",
       "3819695  junction,exon  \n",
       "3819696  junction,exon  \n",
       "3819697                 \n",
       "\n",
       "[3819756 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label_list = [list(ele) for ele in label]\n",
    "label = [','.join(ele) for ele in label_list]\n",
    "bed_df['Label'] = label\n",
    "bed_df['Label'] = bed_df['Label'].replace(r'^\\s*$', 'intron', regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5a3e55ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_df.to_csv('../data/pre-train/rna_selection.bed',sep = '\\t', index=False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d99232e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7639512/7639512 [00:03<00:00, 2057202.43it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "fasta = open('../data/pre-train/rna_seq.fa', 'r')\n",
    "lines = fasta.readlines()\n",
    "seq = []\n",
    "# Strips the newline character\n",
    "for line in tqdm(lines):\n",
    "    if line[0] == '>':\n",
    "        next\n",
    "    else: \n",
    "        seq.append(line.strip().upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2afcae97",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import numpy as np\n",
    "file_name = '../data/pre-train/rna_seq.h5'\n",
    "h5f = h5py.File(file_name, 'w')\n",
    "x = h5f.create_dataset('seq',data = seq)\n",
    "h5f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f83ef64a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9e512872",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_df = pd.read_csv('../data/pre-train/region_selection.bed',sep='\\t',index_col=None)\n",
    "\n",
    "bed_df['Label'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6389e78e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['junction,exon', 'exon', 'intron', 'junction,exon,UTR',\n",
       "       'junction,UTR'], dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "712b3416",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18500</td>\n",
       "      <td>19012</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19012</td>\n",
       "      <td>19524</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19524</td>\n",
       "      <td>20036</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>20036</td>\n",
       "      <td>20548</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>20548</td>\n",
       "      <td>21060</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819741</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57203191</td>\n",
       "      <td>57203703</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819742</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57207481</td>\n",
       "      <td>57207993</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819743</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57207993</td>\n",
       "      <td>57208505</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819747</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210041</td>\n",
       "      <td>57210553</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3819755</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57214232</td>\n",
       "      <td>57214744</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3455744 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Chr     Start       End Strand                      Gene   Label\n",
       "13       chr1     18500     19012      -         ENSG00000227232.5  intron\n",
       "14       chr1     19012     19524      -         ENSG00000227232.5  intron\n",
       "15       chr1     19524     20036      -         ENSG00000227232.5  intron\n",
       "16       chr1     20036     20548      -         ENSG00000227232.5  intron\n",
       "17       chr1     20548     21060      -         ENSG00000227232.5  intron\n",
       "...       ...       ...       ...    ...                       ...     ...\n",
       "3819741  chrY  57203191  57203703      -  ENSG00000185203.12_PAR_Y  intron\n",
       "3819742  chrY  57207481  57207993      +  ENSG00000182484.15_PAR_Y  intron\n",
       "3819743  chrY  57207993  57208505      +  ENSG00000182484.15_PAR_Y  intron\n",
       "3819747  chrY  57210041  57210553      +  ENSG00000182484.15_PAR_Y  intron\n",
       "3819755  chrY  57214232  57214744      -   ENSG00000227159.8_PAR_Y  intron\n",
       "\n",
       "[3455744 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_df[bed_df['Label'].str.contains(\"intron\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d097ef55",
   "metadata": {},
   "source": [
    "Count of reads containing each label\n",
    "\n",
    "'UTR':19953\n",
    "\n",
    "'exon':363682\n",
    "\n",
    "'junction': 284194\n",
    "\n",
    "'intron': 3455744\n",
    "\n",
    "\n",
    "\n",
    "Not balanced. Sub-sample?\n",
    "\n",
    "Total read :3819756"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "154842f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "67"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('CTTCTATTTATTTATTTATTTATTTATTTGTTTGTTTTAGAAGATTCTATGTTAATATTTTATGTGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e974555",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.10 ('tf_2': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "vscode": {
   "interpreter": {
    "hash": "9e8284e1417b754e460c2bde3a4a4837c482fa82ceb7d52f4acbe340dd4b4559"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
